### ISSUE TITLE
Multiply and divide by 0x100000000 ?

### ISSUE THREAD

**Issue by bratao** [hiroshi-manabe/CRFSegmenter#6]
Hello @hiroshi-manabe , how you doing?

I just saw your latest commit, that changes Mutex -> Atomic.

Out of curiosity, you make a copy to the `vector<atomic_uint_least64_t> g2;`
Than multiply by `0x100000000`

Later you divide by `0x100000000`

What is motivation of doing this?

Best regards,
Bruno

**Comment by hiroshi-manabe** [comment]
Hello @bratao, thank you for commenting!

First, I had to use an integer type because [std::atomic&lt;double&gt; is not fully supported](https://stackoverflow.com/questions/30048533/why-isnt-atomic-double-fully-implemented).

As you can probably imagine, emulating a floating point type using an integer one isn't a piece of cake, so I decided to emulate a fixed point number type, multiplying by a certain number when converting from a floating point type and dividing by the same number when re-converting to a floating point type. (A fixed point number type would do the work because the values wouldn't be much far from zero)

I chose a type that has at least 64 bits (int_least64_t), and I decided to use half of the bits (32 bits) as fractional bits, so comes the scaling constant 0x100000000.

I'll consider adding comments on this point.

**Comment by bratao** [comment]
Oh, I understand now.

It is much faster than using a mutex ? 
Did you observed even slightly worse results,  using this integer in place of a double ?

Thank you =)

**Comment by hiroshi-manabe** [comment]
> It is much faster than using a mutex ?

Yes, it's _much_ faster than using a mutex.
Using the old version, the CPU load stopped at 5 % or so, however many threads I used.
(This was Linux specific; It worked fine on Windows. I can't figure out why.)
The new version achieves a constant load of 80 %.

> Did you observed even slightly worse results, using this integer in place of a double ?

No, as far as I observed, and I'm reasonably sure that this change won't affect the results, because 32 bit accuracy below the decimal point is more than sufficient for this application.
